### PR TITLE
Librification

### PR DESCRIPTION
**Pull request by wmanley**
The goal is to make stbt more amenable to fitting into other testing frameworks such as [py.test] or [behave].  Roughly this means reducing and better managing global state.  The hope in the future is that you will be able to run several tests serially in the same process.

Notably:

* The global variables `_display` and `_control` have been replaced by a single global `_dut` of type `DeviceUnderTest`
* All functions that require access to the display or the control have become methods of `DeviceUnderTest`, and access them via `self`
* `stbt/__init__.py` just forwards calls onto it's global `DeviceUnderTest` instance
* `DeviceUnderTest` is a `contextmanager`, so no state is going to be mutated unless `__enter__` has been called, and everything should be cleaned up by `__exit__`

The result is something more functional taking more care over managing state.

Future pull-requests will continue to evolve in this direction:

1. To split the logging aspects (including the `sink_pipeline`) of `DeviceUnderTest` into a separate class, which can be instantiated and used separately.
2. To remove all `get_config` calls from `DeviceUnderTest`, passing them in at construction time instead.

Note: `DeviceUnderTest` is still a private class and this introduces no new public API.

TODO:

- [x] Decide what to do about duplicating documentation.  Possible just remove it from `DeviceUnderTest`

[py.test]: http://pytest.org/
[behave]: https://pypi.python.org/pypi/behave

### CODE REVIEW

**Comment by drothlis** [comment]
> Decide what to do about duplicating documentation. Possible just remove it from `DeviceUnderTest`

I agree, remove it from `DeviceUnderTest`.

File: _stbt/core.py [hunk 1]
```diff
@@ -0,0 +1,2329 @@
...
+                self.region,
+                "%dx%dx%d" % (self.frame.shape[1], self.frame.shape[0],
+                              self.frame.shape[2]),
+                repr(self.text)))
+
+
+def new_device_under_test_from_config(
+        gst_source_pipeline=None, gst_sink_pipeline=None, control_uri=None,
+        save_video=False, restart_source=False,
+        transformation_pipeline='identity'):
```
Comment (drothlis) [inline _stbt/core.py]: Unlike the name suggests, this doesn't actually read in `gst_source_pipeline` etc. from config if it isn't given as an argument.

Comment (wmanley) [inline _stbt/core.py]: Fixed

**Comment by wmanley** [comment]
>> Decide what to do about duplicating documentation. Possible just remove it from DeviceUnderTest

> I agree, remove it from DeviceUnderTest.

Done

File: _stbt/core.py (hunk 1 above)
Comment (drothlis) [inline _stbt/core.py]: How about `run.save_video`, `global.restart_source`, and `global.transformation_pipeline`?

**Comment by drothlis** [comment]
`test_invalid_source_pipeline` fails for me:

    error: The last error message in '/tmp/stb-tester.T1I/stbt.log' wasn't the expected 'no element "viddily-boo"'

**Comment by drothlis** [comment]
After adding more logging:

```
$ tests/run-tests.sh test_invalid_source_pipeline
test_invalid_source_pipeline... FAIL
Showing '/tmp/stb-tester.L3Y/log':
stbt-run: Arguments:
control: test
save_trace: None
verbose: 1
args: []
source_pipeline: viddily-boo
script: test.py
restart_source: False
structured_logging: None
sink_pipeline: fakesink sync=false
save_video: 
stbt-run: source pipeline: viddily-boo ! queue name=_stbt_user_data_queue max-size-buffers=0     max-size-bytes=0 max-size-time=10000000000 ! decodebin ! queue name=_stbt_raw_frames_queue max-size-buffers=2 ! videoconvert ! video/x-raw,format=BGR ! identity ! appsink name=appsink max-buffers=1 drop=false sync=true emit-signals=true caps=video/x-raw,format=BGR
stbt-run: sink pipeline: appsrc name=appsrc format=time caps=video/x-raw,format=(string)BGR ! tee name=t  t. ! queue leaky=downstream ! videoconvert ! fakesink sync=false
FAIL: test.py: GError: no element "viddily-boo"
Traceback (most recent call last):
  File "/tmp/stbt-test-installation.UuNJDV/libexec/stbt/stbt-run", line 71, in <module>
    stbt.get_config('global', 'transformation_pipeline'))
  File "/tmp/stbt-test-installation.UuNJDV/libexec/stbt/stbt/__init__.py", line 434, in init_run
    _dut.__enter__()
  File "/tmp/stbt-test-installation.UuNJDV/libexec/stbt/_stbt/core.py", line 549, in __enter__
    self._display.startup()
  File "/tmp/stbt-test-installation.UuNJDV/libexec/stbt/_stbt/core.py", line 1549, in startup
    self.create_source_pipeline()
  File "/tmp/stbt-test-installation.UuNJDV/libexec/stbt/_stbt/core.py", line 1318, in create_source_pipeline
    self.source_pipeline_description)
GError: no element "viddily-boo"
stbt-run: teardown: Sending eos
Traceback (most recent call last):
  File "/tmp/stbt-test-installation.UuNJDV/libexec/stbt/stbt-run", line 127, in <module>
    stbt.teardown_run()
  File "/tmp/stbt-test-installation.UuNJDV/libexec/stbt/stbt/__init__.py", line 438, in teardown_run
    _dut.__exit__(None, None, None)
  File "/tmp/stbt-test-installation.UuNJDV/libexec/stbt/_stbt/core.py", line 554, in __exit__
    self._display.teardown()
  File "/tmp/stbt-test-installation.UuNJDV/libexec/stbt/_stbt/core.py", line 1570, in teardown
    self.mainloop_thread.join(10)
  File "/usr/lib/python2.7/threading.py", line 938, in join
    raise RuntimeError("cannot join thread before it is started")
RuntimeError: cannot join thread before it is started
```

**Comment by drothlis** [comment]
I've finished my review. Great to see this finally happening!

**Pull request merged**